streaming/src/event_server.cc

```diff
@@ -0,0 +1,96 @@
+#include <unordered_set>
+
+#include "event_server.h"
+namespace ray {
+namespace streaming {
+EventServer::EventServer(uint32_t event_size)
+    : event_queue_(std::make_shared<EventQueue>(event_size)), stop_flag_(false) {}
+EventServer::~EventServer() {
+  stop_flag_ = true;
+  // No need to join if loop thread has never been created.
+  if (loop_thread_ && loop_thread_->joinable()) {
+    STREAMING_LOG(WARNING) << "Loop Thread Stopped";
+    loop_thread_->join();
+  }
+}
+
+void EventServer::Run() {
+  stop_flag_ = false;
+  event_queue_->Run();
+  loop_thread_ = std::make_shared<std::thread>(&EventServer::LoopThreadHandler, this);
+  STREAMING_LOG(WARNING) << "event_server run";
+}
+
+void EventServer::Stop() {
+  stop_flag_ = true;
+  event_queue_->Stop();
+  if (loop_thread_->joinable()) {
+    loop_thread_->join();
+  }
+  STREAMING_LOG(WARNING) << "event_server stop";
+}
+
+bool EventServer::Register(const EventType &type, const Handle &handle) {
+  if (event_handle_map_.find(type) != event_handle_map_.end()) {
+    STREAMING_LOG(WARNING) << "EventType had been registered!";
+  }
+  event_handle_map_[type] = handle;
+  return true;
+}
+
+void EventServer::Push(const Event &event) { event_queue_->Push(event); }
+
+void EventServer::Execute(Event &event) {
+  if (event_handle_map_.find(event.type) == event_handle_map_.end()) {
+    STREAMING_LOG(WARNING) << "Handle has never been registered yet, type => "
+                           << static_cast<int>(event.type);
+    return;
+  }
+  Handle &handle = event_handle_map_[event.type];
+  if (handle(event.channel_info)) {
+    event_queue_->Pop();
+  }
```

Comment:
If handle event return false, the event won't be popped from queue. Is this an expected behavior?

Reply:
> If handle event return false, the event won't be popped from queue. Is this an expected behavior?

Retry to execute again if handle returns false.